vsintegration/src/FSharp.Editor/LanguageService/FSharpProjectOptionsManager.fs

```diff
@@ -334,93 +334,150 @@ type private FSharpProjectOptionsReactor (workspace: Workspace, settings: Editor
+                            return None
+                        else
+                            return options
+                    else
+                        // We only care about the latest project in the workspace's solution.
+                        // We do this to prevent any possible cache thrashing in FCS.
+                        let project = document.Project.Solution.Workspace.CurrentSolution.GetProject(document.Project.Id)
+                        if not (isNull project) then
+                            let! options = tryComputeOptions project ct
+                            if ct.IsCancellationRequested then
+                                return None
+                            else
+                                return options
+                        else
+                            return None
+                with
+                | _ ->
+                    return None
+        }
+
+    let tryGetOptionsByProject (project: Project, ct: CancellationToken) =
+        async {
+            if ct.IsCancellationRequested then
+                return None
+            else
+                try
+                    if project.Solution.Workspace.Kind = WorkspaceKind.MiscellaneousFiles || project.IsFSharpMiscellaneousOrMetadata then
+                        return None
+                    else
+                        // We only care about the latest project in the workspace's solution.
+                        // We do this to prevent any possible cache thrashing in FCS.
+                        let project = project.Solution.Workspace.CurrentSolution.GetProject(project.Id)
+                        if not (isNull project) then
+                            let! options = tryComputeOptions project ct
+                            if ct.IsCancellationRequested then
+                                return None
+                            else
+                                return options
+                        else
+                            return None
+                with
+                | _ ->
+                    return None
+        }
+
+    let clearOptions projectId =
+        match cache.TryRemove(projectId) with
+        | true, (_, _, projectOptions) ->
+            lastSuccessfulCompilations.TryRemove(projectId) |> ignore
+            checkerProvider.Checker.ClearCache([projectOptions])
+        | _ ->
+            ()
+        legacyProjectSites.TryRemove(projectId) |> ignore
+
+    let clearSingleFileOptionsCache documentId =
+        match singleFileCache.TryRemove(documentId) with
+        | true, (_, _, projectOptions) ->
+            lastSuccessfulCompilations.TryRemove(documentId.ProjectId) |> ignore
+            checkerProvider.Checker.ClearCache([projectOptions])
+        | _ ->
+            ()
+
+    [<Literal>]
+    let useMailboxProcessor =
+#if DEBUG
+        false
+#else
+        true
+#endif
+
+    let gate = obj()
+    let asyncLock f =
+        async {
```

Comment:
This is a global lock and asyncLock always uses it.  Which is probably not great.